dwave/system/testing.py

```diff
@@ -108,23 +114,32 @@ class MockDWaveSampler(dimod.Sampler, dimod.Structured):
         -1
 
     """
-    # Feature suggestion - add seed as an optional input, to allow reproducibility.
-
     nodelist = None
     edgelist = None
     properties = None
     parameters = None
-
-    # by default, use ExactSolver for problems up to size (inclusive):
+    
     EXACT_SOLVER_CUTOFF_DEFAULT = 16
 
     def __init__(self,
                  nodelist=None, edgelist=None, properties=None,
                  broken_nodes=None, broken_edges=None,
                  topology_type=None, topology_shape=None,
                  parameter_warnings=True,
+                 substitute_sampler=None, 
+                 substitute_kwargs=None, 
                  exact_solver_cutoff=EXACT_SOLVER_CUTOFF_DEFAULT,
                  **config):
+        
+        self.mocked_parameters={'answer_mode',
+                       'max_answers',
+                       'num_reads',
+                       'label',
+                       'initial_state',
+        }
+
+        self.substitute_sampler = substitute_sampler or SteepestDescentSampler()  # Exact where possible, SteepestDescent otherwise
+        self.substitute_kwargs = substitute_kwargs or {}  # Matched to mocked_parameters
```

Comment:
It's safer to check for `None` identity than to use a shortcircuiting property of logical operators in Python. 

Best practice idiom for setting default mutable values of input params is:
```
if x is None:
    x = <mutable object value>
```
(Emphasis on mutable because immutable can be just set as a kwarg default.)

The reason? See [truth value testing](https://docs.python.org/3/library/stdtypes.html#truth-value-testing) and [`__bool__()`](https://docs.python.org/3/reference/datamodel.html#object.__bool__) method.